File: macclean/cleaners/quicklook.py
from pathlib import Path
import click
from rich.console import Console
from rich.panel import Panel
from macclean.core.utils import AnalysisResult, CleanItem, format_size, confirm, dir_size, run_cmd

console = Console()

_QL_CACHE_DIRS = [
    "com.apple.QuickLookDaemon",
    "com.apple.quicklook.ThumbnailsAgent",
]


def analyze(home: Path | None = None) -> AnalysisResult:
    home = home or Path.home()
    result = AnalysisResult()
    caches = home / "Library" / "Caches"
    for name in _QL_CACHE_DIRS:
        path = caches / name
        if path.exists():
            result.items.append(CleanItem(label=name, path=path, size_bytes=dir_size(path)))
    return result


def clean(result: AnalysisResult, dry_run: bool = False, yes: bool = False) -> None:
    console.print(Panel("[bold cyan]QuickLook Cache Rebuild[/]"))
    total = format_size(result.total_bytes) if result.total_bytes else "0 B"
    console.print(f"  Stale QL cache: [bold]{total}[/]")

    if dry_run:
        return
    if not yes and not confirm("Kill and rebuild QuickLook server?"):
        return

    for args, label in [
        (["qlmanage", "-r"], "Kill QL server"),
        (["qlmanage", "-r", "cache"], "Rebuild QL cache"),
    ]:
        out, code = run_cmd(args)
        if code == 0 or "reset" in out.lower():
            console.print(f"  [green]+[/] {label}")
        else:
            console.print(f"  [yellow]![/] {label}: {out}")


@click.command()
@click.pass_context
def cmd(ctx):
    """Kill and rebuild the QuickLook server and cache."""
    result = analyze()
    clean(result, dry_run=ctx.obj["dry_run"], yes=ctx.obj["yes"])

File: src/cleaners/quicklook.rs
use std::path::PathBuf;
use anyhow::Result;
use crate::core::{AnalysisResult, Cleaner};
use crate::core::fs::dir_size;
use crate::core::cmd::run_cmd;
use crate::ui;

pub struct QuickLookCleaner;

impl Cleaner for QuickLookCleaner {
    fn name(&self) -> &str { "quicklook" }
    fn display_name(&self) -> &str { "QuickLook Cache" }

    fn analyze(&self) -> Result<AnalysisResult> {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
        let mut result = AnalysisResult::default();

        let ql_dirs = [
            ("com.apple.QuickLookDaemon", "Library/Caches/com.apple.QuickLookDaemon"),
            ("com.apple.quicklook.ThumbnailsAgent", "Library/Caches/com.apple.quicklook.ThumbnailsAgent"),
        ];

        for (label, rel) in &ql_dirs {
            let path = home.join(rel);
            if path.exists() {
                let size = dir_size(&path);
                if size > 0 {
                    result.add(*label, path, size);
                }
            }
        }

        Ok(result)
    }

    fn clean(&self, result: &AnalysisResult, dry_run: bool, yes: bool) -> Result<()> {
        if result.items.is_empty() {
            println!("No QuickLook cache found.");
            return Ok(());
        }
        ui::print_analysis("QuickLook Cache", &result.items);
        if dry_run { return Ok(()); }
        if !yes && !ui::confirm("Kill and rebuild QuickLook server?", false)? { return Ok(()); }

        run_cmd(&["qlmanage", "-r"]);
        run_cmd(&["qlmanage", "-r", "cache"]);
        ui::print_ok("QuickLook server reset and cache rebuilt");

        Ok(())
    }
}
